src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs

```diff
@@ -0,0 +1,184 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using static Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles.SymbolSpecification;
+using Words = System.Collections.Generic.IEnumerable<string>;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    internal partial class DeclarationNameCompletionProvider : CommonCompletionProvider
+    {
+        internal override bool IsInsertionTrigger(SourceText text, int insertedCharacterPosition, OptionSet options)
+        {
+            return CompletionUtilities.IsTriggerAfterSpaceOrStartOfWordCharacter(text, insertedCharacterPosition, options);
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+            var position = completionContext.Position;
+            var document = completionContext.Document;
+            var cancellationToken = completionContext.CancellationToken;
+            var semanticModel = await document.GetSemanticModelForSpanAsync(new Text.TextSpan(position, 0), cancellationToken).ConfigureAwait(false);
+
+            var context = CSharpSyntaxContext.CreateContext(document.Project.Solution.Workspace, semanticModel, position, cancellationToken);
+
+            var nameInfo = await NameDeclarationInfo.GetDeclarationInfo(document, position, cancellationToken).ConfigureAwait(false);
+
+            if (!IsValidType(nameInfo.Type))
+            {
+                return;
+            }
+
+            var type = UnwrapType(nameInfo.Type);
+            var baseNames = NameGenerator.GetBaseNames(type);
+            int sortValue = 0;
+            foreach (var (name, kind) in GetRecommendedNames(baseNames, nameInfo, context))
+            {
+                // We've produced items in the desired order, add a sort text to each item to prevent alphabetization
+                completionContext.AddItem(CreateCompletionItem(name, GetGlyph(kind, nameInfo.DeclaredAccessibility), sortValue++.ToString("D8")));
+            }
+
+            AddBuilder(completionContext);
+        }
+
+        private bool IsValidType(ITypeSymbol type)
+        {
+            if (type == null)
+            {
+                return false;
+            }
+
+            if (type.IsErrorType() && (type.Name == "var" || type.Name == string.Empty))
+            {
+                return false;
+            }
+
+            return type.SpecialType != SpecialType.System_Void;
+        }
+
+        private Glyph GetGlyph(SymbolKind kind, Accessibility declaredAccessibility)
+        {
+            Glyph publicIcon;
+            switch (kind)
+            {
+                case SymbolKind.Field:
+                    publicIcon = Glyph.FieldPublic;
```

Comment:
There is a similar extension method that takes an ISymbol. It's hard to extract it out to be useful because it looks at more ISymbol properties than the kind and accessibility.